lib/qemu_sandbox.ml

```diff
@@ -0,0 +1,221 @@
+              @ extra_mounts in
+  let _, proc = Os.open_process ~stdin:qemu_stdin ~stdout:`Dev_null ~pp cmd in
+
+  let ssh = ["ssh"; "opam@localhost"; "-p"; port; "-o"; "NoHostAuthenticationForLocalhost=yes"] in
+
+  let rec loop = function
+    | 0 -> Lwt_result.fail (`Msg "No connection")
+    | n ->
+      Os.exec_result ~pp (ssh @ ["exit"]) >>= function
+      | Ok _ -> Lwt_result.ok (Lwt.return ())
+      | _ -> Lwt_unix.sleep 1. >>= fun _ -> loop (n - 1) in
+  Lwt_unix.sleep 5. >>= fun _ ->
+  loop t.qemu_boot_time >>= fun _ ->
+
+  Lwt_list.iteri_s (fun i { Config.Mount.dst; _ } ->
+    match t.qemu_guest_os with
+    | Linux ->
+      let dev = Printf.sprintf "/dev/vd%c1" (Char.chr (Char.code 'b' + i)) in
+      Os.exec (ssh @ ["sudo"; "mount"; dev; dst])
+    | OpenBSD ->
+      let dev = Printf.sprintf "/dev/sd%ca" (Char.chr (Char.code '1' + i)) in
+      Os.exec (ssh @ ["doas"; "fsck"; "-y"; dev]) >>= fun () ->
+      Os.exec (ssh @ ["doas"; "mount"; dev; dst])
+    | Windows ->
+      Os.exec (ssh @ ["cmd"; "/c"; "rmdir /s /q '" ^ dst ^ "'"]) >>= fun () ->
+      let drive_letter = String.init 1 (fun _ -> Char.chr (Char.code 'd' + i)) in
+      Os.exec (ssh @ ["cmd"; "/c"; "mklink /j '" ^ dst ^ "' '" ^ drive_letter ^ ":\\'"])
+  ) config.Config.mounts >>= fun () ->
+
+  Os.with_pipe_from_child @@ fun ~r:out_r ~w:out_w ->
+  let stdin = Option.map (fun x -> `FD_move_safely x) stdin in
+  let stdout = `FD_move_safely out_w in
+  let stderr = stdout in
+  let copy_log = copy_to_log ~src:out_r ~dst:log in
+  let env = List.map (fun (k, v) -> k ^ "=" ^ v) config.Config.env |> Array.of_list in
+  let sendenv = if Array.length env > 0 then List.map (fun (k, _) -> ["-o"; "SendEnv=" ^ k]) config.Config.env |> List.flatten else [] in
+  let _, proc2 = Os.open_process ~env ?stdin ~stdout ~stderr ~pp (ssh @ sendenv @ ["cd"; config.Config.cwd; "&&"] @ config.Config.argv) in
+  Lwt.on_termination cancelled (fun () ->
+    let aux () =
+      if Lwt.is_sleeping proc then
+        Lwt_io.write qemu_monitor "quit\n"
+      else Lwt.return_unit (* Process has already finished *)
+    in
+    Lwt.async aux
+  );
+  Os.process_result ~pp proc2 >>= fun res ->
+  copy_log >>= fun () ->
+
+  (match t.qemu_guest_arch with
+    | Amd64 ->
+      Log.info (fun f -> f "Sending QEMU an ACPI shutdown event");
+      Lwt_io.write qemu_monitor "system_powerdown\n"
+    | Riscv64 ->
+      (* QEMU RISCV does not support ACPI until >= v9 *)
+      Log.info (fun f -> f "Shutting down the VM");
+      Os.exec (ssh @ ["sudo"; "poweroff"])) >>= fun () ->
+  let rec loop = function
+  | 0 ->
+    Log.warn (fun f -> f "Powering off QEMU");
+    Lwt_io.write qemu_monitor "quit\n"
+  | n ->
+    if Lwt.is_sleeping proc then 
+      Lwt_unix.sleep 1. >>= fun () ->
+      loop (n - 1)
+    else Lwt.return () in
+  loop t.qemu_boot_time >>= fun _ ->
+  
+  Os.process_result ~pp proc >>= fun _ ->
```

Comment:
Looks like we are also ignoring the error result in case the `proc` fails here. IIUC, it would be helpful to report and handle the failure here. But, if not, I think this could just use `Os.exec`, and then the `~pp` wouldn't be needed?

Reply:
`Os.exec` runs a process and waits for the result so couldn't be used.  `proc` is the QEMU process.  I'm starting the process much earlier using `Os.open_process` and using `Os.process_result` to collect the exit status.  We could log the result but ultimately the exit status of `proc2`, which is the test step, is what is important.

Reply:
Ah yes, I see that now, thanks to your explanation and having now worked through all the code. Thanks!